File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/move/StockPileToColumnMove.java
package org.silnith.game.solitaire.move;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.ListIterator;
import java.util.Map;

import org.silnith.deck.Card;
import org.silnith.deck.Suit;
import org.silnith.game.solitaire.Board;
import org.silnith.game.solitaire.Column;
import org.silnith.util.Pair;


/**
 * A move that takes a card from the stock pile and puts it onto a column.
 * 
 * <p>If the column is empty, the card must be a {@link org.silnith.deck.Value#KING}.
 * If the column is not empty, the card must follow the rules of a run.</p>
 */
public class StockPileToColumnMove implements SolitaireMove {
	
	/**
	 * Finds all moves where a card is drawn from the stock pile to a column run.
	 * 
	 * @param board the board to examine
	 * @return a collection of moves
	 */
	public static Collection<StockPileToColumnMove> findMoves(final Board board) {
		if (board.getStockPileIndex() > 0) {
			final Collection<StockPileToColumnMove> moves = new ArrayList<>(7);
			final Card card = board.getStockPileCard();
			final List<Card> run = Collections.singletonList(card);
			final List<Column> columns = board.getColumns();
			final ListIterator<Column> iter = columns.listIterator();
			while (iter.hasNext()) {
				final int index = iter.nextIndex();
				final Column column = iter.next();
				if (column.canAddRun(run)) {
					moves.add(new StockPileToColumnMove(index, board));
					// TODO: Short-circuit if the card is a king?
				}
			}
			return moves;
		}
		return Collections.emptySet();
	}
    
    /**
     * The index into the stock pile from which the card is taken.
     */
    private final int sourceIndex;
    
    /**
     * The index into the board of the column to which the card is being moved.
     */
    private final int destinationColumnIndex;
    
    /**
     * The card being moved.
     */
    private final Card card;
    
    /**
     * Creates a new move of a card from the stock pile to a column on the board.
     * 
     * @param sourceIndex the stock pile index of the card being moved
     * @param destinationColumnIndex the index into the board of the destination column
     * @param card the card being moved
     */
    public StockPileToColumnMove(final int sourceIndex, final int destinationColumnIndex, final Card card) {
        super();
        this.sourceIndex = sourceIndex;
        this.destinationColumnIndex = destinationColumnIndex;
        this.card = card;
    }
    
    /**
     * Creates a new move of a card from the stock pile to a column on the board.
     * 
     * @param destinationColumnIndex the index into the board of the destination column
     * @param board the board from which to get the card
     * @throws IndexOutOfBoundsException if no card is available to be drawn from the board stock pile
     */
    public StockPileToColumnMove(final int destinationColumnIndex, final Board board) {
    	this(board.getStockPileIndex(), destinationColumnIndex, board.getStockPileCard());
    }
    
    /**
     * Returns the index into the stock pile of the card being moved.
     * 
     * @return the stock pile index
     */
    public int getSourceIndex() {
        return sourceIndex;
    }
    
    /**
     * Returns the card being moved.
     * 
     * @return the card being moved
     * @see #getCards()
     */
    public Card getCard() {
        return card;
    }
    
    @Override
    public boolean hasCards() {
        return true;
    }
    
    @Override
    public List<Card> getCards() {
        return Collections.singletonList(card);
    }

    @Override
    public boolean isStockPileModification() {
        return true;
    }

    @Override
    public boolean isStockPileAdvance() {
        return false;
    }

    @Override
    public boolean isStockPileRecycle() {
        return false;
    }

    @Override
    public boolean isFromStockPile() {
        return true;
    }

    @Override
    public boolean isFromFoundation() {
        return false;
    }

    @Override
    public boolean isFromColumn() {
        return false;
    }

    @Override
    public boolean isFromColumn(int columnIndex) {
        return false;
    }

    @Override
    public int getSourceColumnIndex() {
        throw new IllegalStateException("Not a move from a column.");
    }

    @Override
    public boolean isToFoundation() {
        return false;
    }

    @Override
    public boolean isToColumn() {
        return true;
    }

    @Override
    public boolean isToColumn(int columnIndex) {
        return columnIndex == destinationColumnIndex;
    }

    @Override
    public int getDestinationColumnIndex() {
        return destinationColumnIndex;
    }

	@Override
    public Board apply(final Board board) {
    	final Pair<Card, List<Card>> pair = board.extractStockPileCard();
        final Card card = pair.getFirst();
		final List<Card> newStockPile = pair.getSecond();
		
		final int stockPileIndex = board.getStockPileIndex();
		final int newStockPileIndex = stockPileIndex - 1;
		
		final List<Column> columns = board.getColumns();
		final Column column = columns.get(destinationColumnIndex);
		final Column newColumn = column.withCard(card);
		
		final List<Column> newColumns = new ArrayList<>(columns);
		newColumns.set(destinationColumnIndex, newColumn);
		
		final Map<Suit, List<Card>> foundation = board.getFoundation();
		return new Board(newColumns, newStockPile, newStockPileIndex, foundation);
    }
    
    @Override
    public int hashCode() {
        return sourceIndex ^ destinationColumnIndex ^ card.hashCode();
    }
    
    @Override
    public boolean equals(final Object obj) {
        if (obj instanceof StockPileToColumnMove) {
            final StockPileToColumnMove move = (StockPileToColumnMove) obj;
            return sourceIndex == move.sourceIndex && destinationColumnIndex == move.destinationColumnIndex && card.equals(move.card);
        } else {
            return false;
        }
    }
    
    @Override
    public String toString() {
        return "Move " + card + " from stock pile index " + sourceIndex + " to column " + destinationColumnIndex + ".";
    }
    
}

File: cpp/Solitaire/Klondike/silnith/game/solitaire/move/StockPileToColumnMove.cpp
#include <silnith/game/solitaire/move/StockPileToColumnMove.h>

#include <silnith/game/deck/card.h>

using namespace silnith::game::deck;

using namespace std;
using namespace std::literals::string_literals;

namespace silnith::game::solitaire::move
{
    vector<shared_ptr<solitaire_move>> StockPileToColumnMove::find_moves(board const& board)
    {
        if (board.get_stock_pile_index() > 0)
        {
            vector<shared_ptr<solitaire_move>> moves{};
            card stock_pile_card{ board.get_stock_pile_card() };
            vector<card> run{};
            run.emplace_back(stock_pile_card);
            for (size_t index{ 0 }; index < board::num_columns; index++)
            {
                if (board.get_column(index).can_add_run(run))
                {
                    moves.emplace_back(
                        static_cast<shared_ptr<solitaire_move>>(
                            make_shared<StockPileToColumnMove>(index, board)));
                }
            }
            return moves;
        }
        else
        {
            return vector<shared_ptr<solitaire_move>>{};
        }
    }

    StockPileToColumnMove::StockPileToColumnMove(size_t source_index, size_t destination_column_index, card const& card)
        : source_index{ source_index },
        destination_column_index{ destination_column_index },
        _card{ card }
    {}

    StockPileToColumnMove::StockPileToColumnMove(size_t destination_column_index, board const& board)
        : StockPileToColumnMove(board.get_stock_pile_index(), destination_column_index, board.get_stock_pile_card())
    {}

    StockPileToColumnMove::StockPileToColumnMove(size_t destination_column_index, shared_ptr<board> const& board)
        : StockPileToColumnMove(destination_column_index, *board)
    {}

    size_t StockPileToColumnMove::get_source_index(void) const
    {
        return source_index;
    }

    bool StockPileToColumnMove::has_cards(void) const
    {
        return true;
    }

    vector<card> StockPileToColumnMove::get_cards(void) const
    {
        vector<card> cards{};
        cards.emplace_back(_card);
        return cards;
    }

    bool StockPileToColumnMove::is_stock_pile_modification(void) const
    {
        return false;
    }

    bool StockPileToColumnMove::is_stock_pile_advance(void) const
    {
        return false;
    }

    bool StockPileToColumnMove::is_stock_pile_recycle(void) const
    {
        return false;
    }

    bool StockPileToColumnMove::is_from_stock_pile(void) const
    {
        return true;
    }

    bool StockPileToColumnMove::is_from_foundation(void) const
    {
        return false;
    }

    bool StockPileToColumnMove::is_from_column(void) const
    {
        return false;
    }

    bool StockPileToColumnMove::is_from_column(std::size_t column_index) const
    {
        return false;
    }

    size_t StockPileToColumnMove::get_source_column_index(void) const
    {
        throw invalid_argument("Not a move from a column.");
    }

    bool StockPileToColumnMove::is_to_foundation(void) const
    {
        return false;
    }

    bool StockPileToColumnMove::is_to_column(void) const
    {
        return true;
    }

    bool StockPileToColumnMove::is_to_column(size_t column_index) const
    {
        return column_index == destination_column_index;
    }

    size_t StockPileToColumnMove::get_destination_column_index(void) const
    {
        return destination_column_index;
    }

    shared_ptr<board> StockPileToColumnMove::apply(shared_ptr<board> const& b) const
    {
        pair<card, vector<card>> pair{ b->extract_stock_pile_card() };
        card drawn_card{ pair.first };
        vector<card> new_stock_pile{ pair.second };

        size_t stock_pile_index{ b->get_stock_pile_index() };
        size_t new_stock_pile_index{ stock_pile_index - 1 };

        vector<column> columns{ b->get_columns() };
        column old_column{ columns.at(destination_column_index) };
        column new_column{ old_column.with_card(drawn_card) };

        vector<column> new_columns{};
        for (size_t index{ 0 }; index < board::num_columns; index++)
        {
            if (index == destination_column_index)
            {
                new_columns.emplace_back(new_column);
            }
            else
            {
                new_columns.emplace_back(columns.at(index));
            }
        }

        return make_shared<board>(
            new_columns,
            new_stock_pile,
            new_stock_pile_index,
            b->get_foundation());
    }

    bool StockPileToColumnMove::operator==(StockPileToColumnMove const& other) const
    {
        return source_index == other.source_index
            && destination_column_index == other.destination_column_index
            && _card == other._card;
    }

    string to_string(StockPileToColumnMove const& move)
    {
        return "Move "s
            + to_string(move.get_cards().front())
            + " from stock pile index "s
            + std::to_string(move.get_source_index())
            + " to column "s
            + std::to_string(move.get_destination_column_index())
            + "."s;
    }

    wstring to_wstring(StockPileToColumnMove const& move)
    {
        return L"Move "s
            + to_wstring(move.get_cards().front())
            + L" from stock pile index "s
            + std::to_wstring(move.get_source_index())
            + L" to column "s
            + std::to_wstring(move.get_destination_column_index())
            + L"."s;
    }

    ostream& operator<<(ostream& out, StockPileToColumnMove const& move)
    {
        out << to_string(move);
        return out;
    }

    wostream& operator<<(wostream& out, StockPileToColumnMove const& move)
    {
        out << to_wstring(move);
        return out;
    }
}
